teach/concepts/data-validation/versions.md

```diff
@@ -0,0 +1,136 @@
+    - **Methods**:
+        - `validate(dataInstance): Promise<ValidationError | null>`:
+            1. **Override** the base class method.
+            2. Evaluate the `criteria` expression against the instance object.
+            3. Return a `ValidationError` if the expression is false, using the rule's details.
+            4. Return `null` if the expression evaluates to true.
+    - **Pseudo-code for the subclass**
+
+        ```js
+        class ExpressionValidationRule extends ValidationRule {
+            constructor(name, message, dependentProperties, criteria) {
+                super(name, message, dependentProperties);
+                this.criteria = criteria;
+            }
+
+            async validate(dataInstance) {
+                // delegate logic to FRB...
+                const isValid = evaluateExpression(this.criteria, dataInstance);
+
+                if (!isValid) {
+                    // Rule failed, return the error object.
+                    return new ValidationError(
+                      message: this.message, // Later, we should use a localization service.
+                      dependentProperties: this.dependentProperties,
+                      rule: this,
+                    );
+                }
+
+                // Rule passed.
+                return null;
+           }
+        }
+        ```
+
+3. `ValidationService`
+
+    - **Purpose**: This service orchestrates the validation process by triggering rules and updating the `invalidityState` property on data objects.
+    - **Methods**:
+        - `validateDataObject(instance, allRules)`:
+            1. **Run All Rules**: Asynchronously execute the `validate` method for each rule in `allRules` with the `instance`.
+            2. **Collect Results**: Gather all `ValidationError` objects from failed validations.
+            3. **Update State**: Replace `instance.invalidityState` with the new error array, if any.
+
+4. `PropertyField` Component Enhancements
+
+    In Version 1, the `PropertyField` component will aonly accept data context through explicit properties since automatic discovery will be implemented later.
+
+    - **New Properties:**
+        - `dataInstances: object[]`: A collection of data instances for editing. Enables a component to modify the same property across multiple objects (though most use cases will involve a single object).
+        - `dataInstance: object` : A convenience getter/setter that simplifies access to to `dataInstances[0]`.
+        - `dataType: ObjectDescriptor`: The object descriptor that defines the schema for the `dataInstance`.
+        - `dataTypeProperty: PropertyDescriptor`: The property descriptor within the `dataType` that this field edits (e.g., the `firstName` property descriptor).
+        - `isTouched: boolean`: Initialized to `false`. Indicates whether the field has been interacted with by the user.
```

Comment:
isEdited? "hasBeenEdited" We need to find better than "isTouched" - Is that reset at some point? Like after a save?

Reply:
Can you elaborate on why you think the PropertyField needs to know whether the field has been interacted with by the user?